packages/backend/src/open_payments/auth/middleware.ts

```diff
@@ -55,26 +53,18 @@ export function createAuthMiddleware({
           ctx.throw(401, `Invalid signature`)
         }
       }
-      await GrantReference.transaction(async (trx: Transaction) => {
-        const grantRef = await grantReferenceService.get(grant.grant, trx)
-        if (grantRef) {
-          if (grantRef.clientId !== grant.clientId) {
-            logger.debug(
-              `clientID ${grant.clientId} for grant ${grant.grant} does not match internal reference clientId ${grantRef.clientId}.`
-            )
-            ctx.throw(500)
-          }
-        } else if (action === AccessAction.Create) {
-          // Grant and client ID's are only stored for create routes
-          await grantReferenceService.create(
-            {
-              id: grant.grant,
-              clientId: grant.clientId
-            },
-            trx
-          )
-        }
-      })
+
+      try {
+        await grantReferenceService.getOrCreate(
+          { id: grant.grant, clientId: grant.clientId },
+          action
+        )
+      } catch (e) {
+        const errInfo = e && typeof e === 'object' && e.stack ? e.stack : e
+        logger.debug(errInfo)
```

Comment:
```suggestion
      } catch (error) {
        const errorMessage = error?.message ?? 'Failed to get or create grant reference'
        logger.error({ error }, errorMessage)
```
would prefer something like this, without needing to check the type etc